### DISCUSSION TITLE
Introducing: The Forum Helpers Dashboard

### DISCUSSION THREAD

**Discussion by Accio1** [theforumhelpers/theforumhelpers.github.io#295] [Announcements]
### Problem
When the 3.0 studios update was released, the restriction was added that managers would no longer be able to edit the description of studios. This caused The Forum Helpers application process to be removed from the studio description and added to the website instead, so it could be edited more easily in the future. However, this means that updating the codeword or application status requires a pull request to be submitted and approved, which is not efficient.

### Solution
Introducing: [The Forum Helpers Dashboard](https://theforumhelpers.herokuapp.com/), a website where Forum Helpers can easily log in and perform actions such as opening or closing applications, changing the codeword, or updating their biography. 

### How to use
To use The Forum Helpers Dashboard, visit: https://theforumhelpers.herokuapp.com/ and then type your Scratch Username into the username field. Press submit and then copy-paste the verification code into [this](https://scratch.mit.edu/studios/30202371/comments) studio. As of right now, this only works for managers (see below for why). You should be signed in, and presented with your dashboard. You have the option to sign out, and see your biography that is currently used on the site. You also have the option to change the codeword and open/close applications. **Note:** as of #294 being merged, this is live which means your changes will be reflected on the actual site.

### Notes
- The Forum Helpers Dashboard is currently in early stages, meaning that the only options available are opening or closing applications, and updating the codeword. In the future, additional features (such as changing your biography) will be added.
- For this reason, it is currently only open to managers. This is because it is traditionally managers who would have editing rights to the description, and the ability to change the codeword and application status.

### Contributing
For bug reports or feature requests for TFH Dashboard, please open an issue in this repository. It will be labeled with the "type: dashboard" label to show that it related to TFH Dashboard. **Important**: For security issues related to TFH Dashboard, please leave a comment on one of my older Scratch projects with your concerns.

Please leave your questions, comments, or concerns related to this dashboard below or in The Forum Helpers studio.

**Comment by CST1229** [comment]
Who coded it?

**Reply by Accio1** [reply]
@CST1229 I did

**Comment by Shluffy** [comment]
Maybe Use a Scratch Account instead?

**Reply by Accio1** [reply]
@Shluffy use a Scratch account for what?

**Reply by CST1229** [reply]
Probably for authentication, and Shluffy probably means logging in with your Scratch username and password, which has obivious security issues

**Reply by penguinmoose** [reply]
Right, doesn't the current auth work fine?

**Comment by jeffalo** [comment]
this is very cool! is it open source? also i can't pass the login flow, it says i'm not a manager.

**Reply by Accio1** [reply]
@jeffalo It isn't open source at the moment, but it will be in the future. Are you typing your username exactly as it is on Scratch? You are on the list of managers that I made so it should work.

Edit: On TFH site, your username was inputted in lowercase (and that was the one I used), but your actual username has a capital at the start. The input is case sensitive so I fixed it and it should be working now.

**Reply by jeffalo** [reply]
thanks!

**Reply by Accio1** [reply]
@jeffalo You're welcome! 

Turns out I forgot to setup a testing database for when I am developing it, so it will be down for maintenance in a half an hour while I make a copy.

**Comment by aidan-crockett** [comment]
why does it have to use php

**Reply by aidan-crockett** [reply]
I guess php is best for server side stuff?

**Reply by Accio1** [reply]
PHP is a server side language so that is why I used it. It would be impossible to do it with just vanilla HTML, CSS, and JS. I could have used a different server side language like NodeJS but PHP seemed like the better option.

**Comment by penguinmoose** [comment]
Nice! I really like the idea. I can't wait for it to be out!

**Reply by Accio1** [reply]
I'm excited for the next update to release! Would you like an invite to TFH organization on GitHub since you are a member of TFH?

**Reply by penguinmoose** [reply]
Sure! Yes! I’ll try to help by finding bugs and making suggestions, though I also know quite some web programming (especially CSS).

**Reply by Accio1** [reply]
Alright, can you please leave a comment on my Scratch profile with your Scratch account just to verify that this account is yours?

**Reply by penguinmoose** [reply]
Ok, did that just now

**Comment by penguinmoose** [comment]
Just saying, it says I’m not a manager. Does it actually get the managers through the Scratch API, or does it use an outdated list?

**Reply by Accio1** [reply]
Right now the managers list is hardcoded so that is why it is outdated. This will be resolved in the next update when it uses a database instead.

**Reply by penguinmoose** [reply]
Ok, got it. Also, would it be easier to make the request to get the list of managers on the frontend webpage, not using a server? I'm pretty sure that you can just make a Scratch API request using just JavaScript.

**Reply by Accio1** [reply]
The managers list is in a database so that it can also store people's bios to be displayed on the site.

**Reply by leahcimto** [reply]
Also, we would rather not spam the Scratch API with requests because I'm pretty sure we would also have to make a GET request to get the user's user ID so we could get their profile picture.

**Reply by penguinmoose** [reply]
> Also, we would rather not spam the Scratch API with requests because I'm pretty sure we would also have to make a GET request to get the user's user ID so we could get their profile picture.

True, so I guess the database will update every now or then automatically?

**Reply by penguinmoose** [reply]
> The managers list is in a database so that it can also store people's bios to be displayed on the site.

So the database has everything? That makes more sense now.

**Reply by Accio1** [reply]
> > Also, we would rather not spam the Scratch API with requests because I'm pretty sure we would also have to make a GET request to get the user's user ID so we could get their profile picture.
> 
> 
> 
> True, so I guess the database will update every now or then automatically?

It will be updated by people who are admins on the dashboard (leahcimto and I). When a new member is added, we will set up their profile in the admin dashboard and then they can perform actions such as editing their bio on their dashboard.

**Reply by penguinmoose** [reply]
> > > Also, we would rather not spam the Scratch API with requests because I'm pretty sure we would also have to make a GET request to get the user's user ID so we could get their profile picture.
> > 
> > 
> > True, so I guess the database will update every now or then automatically?
> 
> It will be updated by people who are admins on the dashboard (leahcimto and I). When a new member is added, we will set up their profile in the admin dashboard and then they can perform actions such as editing their bio on their dashboard.

But why not automatically? It can do it one piece of data at a time (with lots of delay like one minute) so it doesn't spam the API. But I guess that would be even harder to do than manually updating it.

**Reply by Accio1** [reply]
Just because the Scratch Team prefers that minimal requests are made to their servers, and the setup for admins to add new users has already been implemented.

**Reply by penguinmoose** [reply]
Oh, ok. That works as well (it'll be on the dashboard thing, right?).

**Reply by leahcimto** [reply]
Yes, admins will have a way to add/remove/edit users and give them access to the dashboard themselves by adding them to the list.

**Reply by penguinmoose** [reply]
Ok nice (I'm going to go off github for now).

**Reply by penguinmoose** [reply]
That is, until tomorrow EDIT: Actually, I did return some time later

**Reply by Accio1** [reply]
@penguinmoose thanks for your engagement. I'm on summer holidays so I actually have time to work on this and I'm glad that people are interacting with the repository again!

**Reply by penguinmoose** [reply]
@Accio1 Me too. I'm actually more active in the summer (unlike some other people).

**Reply by penguinmoose** [reply]
> Yes, admins will have a way to add/remove/edit users and give them access to the dashboard themselves by adding them to the list.

But then how will new managers be able to add themselves to the list, or will someone need to add them?

**Reply by Accio1** [reply]
> > Yes, admins will have a way to add/remove/edit users and give them access to the dashboard themselves by adding them to the list.
> 
> 
> 
> But then how will new managers be able to add themselves to the list, or will someone need to add them?

Admins are responsible for adding all new users.

**Reply by penguinmoose** [reply]
> > > Yes, admins will have a way to add/remove/edit users and give them access to the dashboard themselves by adding them to the list.
> > 
> > 
> > But then how will new managers be able to add themselves to the list, or will someone need to add them?
> 
> Admins are responsible for adding all new users.

You mean studio managers, or something else? What does “admin” mean in this case? Does it have to do with people that are on this GitHub thing?

**Reply by leahcimto** [reply]
Admin so far means @Accio1 and I. I assume it's because we are the two that are most active here and have write access.

**Reply by penguinmoose** [reply]
> Admin so far means @Accio1 and I. I assume it's because we are the two that are most active here and have write access.

I see.

**Comment by penguinmoose** [comment]
I’m thinking that maybe someone can even apply directly on the website, though that would pose lots of security and spam issues.

**Comment by penguinmoose** [comment]
Can someone add me to the list? (I'm helloworldbyeworld)

**Reply by leahcimto** [reply]
See #303. Your name is on the list, and it will be pushed later this week with the dashboard :)

**Reply by penguinmoose** [reply]
Thanks :)
